### PR TITLE
Use hashicorp/memberlist to speedup dead node detection

### LINKED ISSUE #298: Failover time very high in layer2 mode
**Is this a bug report or a feature request?**:

Both, probably.

**What happened**:

I tested with layer 2 mode and simulated node failure by shutting down the node that the load balancer IP was on. What then happened it that it took approx. 5 minutes for the IP to be switched to the other node in the cluster (1 master, 2 nodes). After much experimentation I came to the conclusion that the *node* being down and being "NotReady" did not initiate the switch of the IP address. The 5 minute timeout seems to be caused by the default pod eviction timeout of Kubernetes, which is 5 minutes. That means it takes 5 minutes for a pod on a node that is not available to be deleted. Default "node monitor grace period is 40 seconds, btw.". So that means it currently takes almost 6 minutes with default confguration for an IP address to be switched.

I made things a lot better by decreasing both settings like this:
    - --pod-eviction-timeout=20s
    - --node-monitor-grace-period=20s
in /etc/kubernetes/manifests/kube-controller-manager.yaml

This makes MetalLB switch the IP in case of node failure in the sub-minute range.

**What you expected to happen**:

To be honest what I would expect that the whole process takes maybe max. 5 seconds.

**How to reproduce it (as minimally and precisely as possible)**:

Create a Kubernetes 1.11.1 cluster with kubeadm (single master, two nodes). Calico networking.

kubectl apply -f https://raw.githubusercontent.com/google/metallb/v0.7.2/manifests/metallb.yaml
kubectl apply -f metallb-cfg.yml
kubectl apply -f tutorial-2.yaml

➜  metallb-test cat metallb-cfg.yml 
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metallb-system
  name: config
data:
  config: |
    address-pools:
    - name: default
      protocol: layer2
      addresses:
      - 10.115.195.206-10.115.195.208

Then

watch curl --connect-timeout 1 http://10.115.195.206

to see if the nginx app is reachable.

Then 

kubectl logs -f --namespace metallb-system speaker-xxxxxxxxx

To see which node has the IP address assigned at the moment.
ssh into the machine and "poweroff".

Wait for how long it takes until the "watch curl" is successful again.

**Anything else we need to know?**:

**Environment**:
- MetalLB version: v0.7.2
- Kubernetes version: v1.11.1
- BGP router type/version: N/A
- OS (e.g. from /etc/os-release): CentOS 7
- Kernel (e.g. `uname -a`):Linux cp-k8s-ghdev02-node-01.ewslab.eos.lcl 3.10.0-693.el7.x86_64 #1 SMP Tue Aug 22 21:09:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by champtar**
This is an early POC to fix #298.
It needs a lot of cleanup, but without tuning I get failover around 3s \o/

### CODE REVIEW

**Comment by schaze** [comment]
Great idea! I really think using Kubernetes to do the failover handling is not a practical approach. Do you have any indication how this will scale with e.g. 50 or 80 nodes?

**Comment by champtar** [comment]
According to this presentation https://www.hashicorp.com/resources/everybody-talks-gossip-serf-memberlist-raft-swim-hashicorp-consul, hashicorp goal is/was 5k nodes, so even if we don't tune memberlist configuration properly with 100 nodes I think we should be fine.

**Comment by champtar** [comment]
@danderson when you have time to review, I don't mind some guidance how you would see this code ;)

**Comment by danderson** [comment]
TL;DR: I love it. But I'm trying to onboard more maintainers aside from me, and @daxmc99 wanted to do the review on this, so I'm going to let them do the first pass.

Thank you for making this!

**Comment by champtar** [comment]
@daxmc99 I've improved the PR a bit, I just need need to fix MemberList logs to use MetalLB logger, but please review when you can

**Comment by champtar** [comment]
I've fixed the logs, here an example:
```
{"branch":"","caller":"main.go:80","commit":"","msg":"MetalLB speaker starting (no version or build info)","ts":"2020-02-26T23:30:08.353241621Z","version":""}
{"caller":"main.go:165","msg":"Node event","node addr":"10.10.52.141","node event":0,"node name":"etienne-ks141","ts":"2020-02-26T23:30:08.360848514Z"}
{"caller":"main.go:166","msg":"Call Force Sync","ts":"2020-02-26T23:30:08.361032071Z"}
{"caller":"announcer.go:103","event":"createARPResponder","interface":"eth0","msg":"created ARP responder for interface","ts":"2020-02-26T23:30:08.362216201Z"}
{"caller":"announcer.go:112","event":"createNDPResponder","interface":"eth0","msg":"created NDP responder for interface","ts":"2020-02-26T23:30:08.362750429Z"}
{"caller":"announcer.go:103","event":"createARPResponder","interface":"cali108c69a7a5b","msg":"created ARP responder for interface","ts":"2020-02-26T23:30:08.363544109Z"}
{"caller":"announcer.go:112","event":"createNDPResponder","interface":"cali108c69a7a5b","msg":"created NDP responder for interface","ts":"2020-02-26T23:30:08.363721254Z"}
{"caller":"announcer.go:103","event":"createARPResponder","interface":"cali62f82a7db1b","msg":"created ARP responder for interface","ts":"2020-02-26T23:30:08.364336672Z"}
{"caller":"announcer.go:112","event":"createNDPResponder","interface":"cali62f82a7db1b","msg":"created NDP responder for interface","ts":"2020-02-26T23:30:08.364634205Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:785: [DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.144:7946","ts":"2020-02-26T23:30:08.403798539Z"}
{"caller":"main.go:165","msg":"Node event","node addr":"10.10.52.142","node event":0,"node name":"etienne-ks142","ts":"2020-02-26T23:30:08.405691549Z"}
{"caller":"main.go:166","msg":"Call Force Sync","ts":"2020-02-26T23:30:08.40577009Z"}
{"caller":"main.go:165","msg":"Node event","node addr":"10.10.52.144","node event":0,"node name":"etienne-ks144","ts":"2020-02-26T23:30:08.405783769Z"}
{"caller":"main.go:166","msg":"Call Force Sync","ts":"2020-02-26T23:30:08.40579436Z"}
{"caller":"main.go:165","msg":"Node event","node addr":"10.10.52.143","node event":0,"node name":"etienne-ks143","ts":"2020-02-26T23:30:08.405802063Z"}
{"caller":"main.go:166","msg":"Call Force Sync","ts":"2020-02-26T23:30:08.40581302Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:785: [DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.141:7946","ts":"2020-02-26T23:30:08.405850847Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:210: [DEBUG] memberlist: Stream connection from=10.10.52.141:43598","ts":"2020-02-26T23:30:08.406190368Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:785: [DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.142:7946","ts":"2020-02-26T23:30:08.407089632Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:785: [DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.143:7946","ts":"2020-02-26T23:30:08.444129524Z"}
{"Memberlist nb join":4,"caller":"main.go:150","error ?":null,"ts":"2020-02-26T23:30:08.446427959Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"default/kubernetes","ts":"2020-02-26T23:30:08.74665996Z"}
{"caller":"main.go:247","event":"endUpdate","msg":"end of service update","service":"default/kubernetes","ts":"2020-02-26T23:30:08.746759896Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"kube-system/coredns","ts":"2020-02-26T23:30:08.746778824Z"}
{"caller":"main.go:247","event":"endUpdate","msg":"end of service update","service":"kube-system/coredns","ts":"2020-02-26T23:30:08.746789656Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"kube-system/kubernetes-dashboard","ts":"2020-02-26T23:30:08.746803624Z"}
{"caller":"main.go:247","event":"endUpdate","msg":"end of service update","service":"kube-system/kubernetes-dashboard","ts":"2020-02-26T23:30:08.746813531Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"test-echo/test-lb","ts":"2020-02-26T23:30:08.746824139Z"}
{"caller":"main.go:251","event":"noConfig","msg":"not processing, still waiting for config","service":"test-echo/test-lb","ts":"2020-02-26T23:30:08.746833965Z"}
{"caller":"main.go:252","event":"endUpdate","msg":"end of service update","service":"test-echo/test-lb","ts":"2020-02-26T23:30:08.74684448Z"}
{"caller":"main.go:356","configmap":"metallb-system/config","event":"startUpdate","msg":"start of config update","ts":"2020-02-26T23:30:08.748325546Z"}
{"caller":"main.go:380","configmap":"metallb-system/config","event":"endUpdate","msg":"end of config update","ts":"2020-02-26T23:30:08.748500961Z"}
{"caller":"k8s.go:395","configmap":"metallb-system/config","event":"configLoaded","msg":"config (re)loaded","ts":"2020-02-26T23:30:08.748526633Z"}
{"caller":"bgp_controller.go:285","event":"nodeLabelsChanged","msg":"Node labels changed, resyncing BGP peers","ts":"2020-02-26T23:30:08.748574971Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"test-echo/test-lb","ts":"2020-02-26T23:30:08.753693673Z"}
{"caller":"main.go:294","event":"endUpdate","msg":"end of service update","service":"test-echo/test-lb","ts":"2020-02-26T23:30:08.753772423Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"default/kubernetes","ts":"2020-02-26T23:30:08.753792833Z"}
{"caller":"main.go:247","event":"endUpdate","msg":"end of service update","service":"default/kubernetes","ts":"2020-02-26T23:30:08.753803373Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"kube-system/coredns","ts":"2020-02-26T23:30:08.753818595Z"}
{"caller":"main.go:247","event":"endUpdate","msg":"end of service update","service":"kube-system/coredns","ts":"2020-02-26T23:30:08.753829001Z"}
{"caller":"main.go:243","event":"startUpdate","msg":"start of service update","service":"kube-system/kubernetes-dashboard","ts":"2020-02-26T23:30:08.753839864Z"}
{"caller":"main.go:247","event":"endUpdate","msg":"end of service update","service":"kube-system/kubernetes-dashboard","ts":"2020-02-26T23:30:08.753849696Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:210: [DEBUG] memberlist: Stream connection from=10.10.52.144:59186","ts":"2020-02-26T23:30:51.954839672Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:785: [DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.144:7946","ts":"2020-02-26T23:30:57.854936649Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:210: [DEBUG] memberlist: Stream connection from=10.10.52.144:59290","ts":"2020-02-26T23:31:21.957591719Z"}
{"caller":"main.go:178","component":"MemberList","msg":"net.go:785: [DEBUG] memberlist: Initiating push/pull sync with: 10.10.52.144:7946","ts":"2020-02-26T23:31:27.857257754Z"}
```

**Comment by champtar** [comment]
Ok I think this is ready

File: manifests/metallb.yaml [hunk 1]
```diff
@@ -296,7 +338,7 @@ spec:
       nodeSelector:
         beta.kubernetes.io/os: linux
       serviceAccountName: speaker
-      terminationGracePeriodSeconds: 0
+      terminationGracePeriodSeconds: 2
```
Comment (daxmc99) [inline manifests/metallb.yaml]: Curious why the change here?

File: speaker/main.go [hunk 2]
```diff
@@ -15,10 +15,15 @@
 package main
 
 import (
+	"crypto/sha256"
 	"flag"
 	"fmt"
+	golog "log"
```
Comment (daxmc99) [inline speaker/main.go]: Is there a reason we need to use the std lib log instead of `gokitlog`

File: manifests/metallb.yaml (hunk 1 above)
Comment (champtar) [reply manifests/metallb.yaml]: It's to give it the time to exit properly / leave the cluster
I chose 1sec max to leave the cluster so 2s for the clean shutdown

File: speaker/main.go (hunk 2 above)
Comment (champtar) [reply speaker/main.go]: For the logs, MemberList accept either an io.Writer or a log.Logger. I prefer the log.Logger so can avoid having the date 2 times, and add the filename:line.
In the end we pass a golog.Logger, with a special io.Writer that use gokitlog (please tell me if we can had some more levels :P)

**Comment by champtar** [comment]
@daxmc99 no problem, I just realised I haven't fixed the tests, I'll try to so that before this weekend

**Comment by champtar** [comment]
I haven't improved/added tests but I've fixed it

**Comment by champtar** [comment]
@mskrocki I've not even run tested the new controller in my testing as almost nothing changes
Can you check if the binary was built with "CGO_ENABLED=0" (check with ldd)
Else can you share your containers ?

**Comment by mskrocki** [comment]
I just created controller from HEAD and I am running into same issue, so it is my env. Sorry for the confusion.

**Comment by champtar** [comment]
@mskrocki no problem. Do look if you built a fully static go binary or not as I remember having similar issue (you compile and link on a glibc system, then copy in a musl container)

**Comment by champtar** [comment]
@danderson: @daxmc99 approved so tell me if you want some changes

File: speaker/main.go [hunk 3]
```diff
@@ -110,14 +147,62 @@ func main() {
...
+	if err := client.Run(stopCh); err != nil {
 		logger.Log("op", "startup", "error", err, "msg", "failed to run k8s client")
 	}
 }
 
+func watchMemberListEvents(logger gokitlog.Logger, eventCh chan memberlist.NodeEvent, stopCh chan struct{}, client *k8s.Client) {
+	for {
+		select {
+		case e := <-eventCh:
+			logger.Log("msg", "Node event", "node addr", e.Node.Addr, "node name", e.Node.Name, "node event", e.Event)
```
Comment (mskrocki) [inline speaker/main.go]: Any chance we could translate the e.Event to more readable value? Currently it is just an ID of an enum.

**Comment by danderson** [comment]
The only thing I would like to change (not necessarily in this PR), is the static encryption key in the manifest. I have two problems with this:
 - ~Nobody reads the manifest before applying it, so my expectation is that ~all MetalLB deployments will run with the same "secretkeytobechanged" key.
 - memberlist supports online key rotation, and I would love to use that to rotate keys autonomously.

In general, I think a good way to do this would be:
 - give controller permissions to manage a k8s Secret, and give speakers permissions to read the secret.
 - Controller automatically generates random keys and updates the Secret periodically. Speakers see this change, and trigger a key rotation via the memberlist API.

That way, the security aspect of memberlist would Just Work, with no configuration required.

But that said, this is a huge improvement to failover time, and I don't want to block releasing it on these changes. So instead, can you tweak the installation docs (in the website subdir) to mention that operators should change the manifest secret key before deploying?

File: speaker/main.go (hunk 3 above)
Comment (champtar) [reply speaker/main.go]: done

**Comment by champtar** [comment]
@danderson I 100% agree on the just work part, I wanted to put the guid of the speaker DaemonSet via the downward API but it's not supported.
Looking at it once more key rotation is indeed supported but it was not clear to me at all on my first read of the docs.
My only concerns with online key rotation is on a system with unstable API server (temporary outage) the key rotation could just make the whole system worse, but I think I'll implement it at some point.

**Comment by champtar** [comment]
Just need to update the kustomize part now

**Comment by champtar** [comment]
Haven't tested kustomize part, can someone tell me if that seems ok ?

**Comment by champtar** [comment]
This worked
```
namespace: metallb-system

resources:
  - metallb.yaml

configMapGenerator:
- name: config
  files:
    - configs/config

secretGenerator:
- name: memberlist
  files:
    - configs/secretkey

generatorOptions:
 disableNameSuffixHash: true
```
@danderson I think this should be good now

**Comment by champtar** [comment]
@danderson friendly ping :)
If you don't have time to do any testing maybe just release a beta/rc so I can get more feedback from users

**Comment by champtar** [comment]
@danderson @daxmc99 pretty please :)

**Comment by danderson** [comment]
Merged, thank you! If @daxmc99 has other feedback, we can do it in a followup change.

We've been talking about scheduling a release in #metallb-dev on the k8s slack, if you want to drop in there.
